XiEditor/StatusBar.swift

```diff
@@ -0,0 +1,140 @@
+//
+//  StatusBar.swift
+//  XiEditor
+//
+//  Created by Dzũng Lê on 19/05/2018.
+//  Copyright © 2018 Raph Levien. All rights reserved.
+//
+
+import Foundation
+import Cocoa
+
+enum StatusItemAlignment: String {
+    case left = "left"
+    case right = "right"
+}
+
+class StatusItem: NSTextField {
+
+    var key: String = ""
+    var value: String = ""
+    var barAlignment: StatusItemAlignment
+
+    init(_ key: String, _ value: String, _ barAlignment: String) {
+        self.key = key
+        self.value = value
+        self.barAlignment = StatusItemAlignment(rawValue: barAlignment)!
+        super.init(frame: NSZeroRect)
+
+        // Similar to what NSTextField convenience init creates
+        self.isEditable = false
+        self.isSelectable = false
+        self.textColor = NSColor.labelColor
+        self.backgroundColor = NSColor.clear
+        self.lineBreakMode = .byClipping
+        self.isBezeled = false
+        self.stringValue = value
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
+
+class StatusBar: NSView {
+
+    private let backgroundColor = NSColor(deviceWhite: 0.9, alpha: 1.0)
+    private let statusBarHeight: CGFloat = 20
+
+    // This should change in the future, as ordering is alphabetical.
+    var leftItems = [StatusItem]()
+    var rightItems = [StatusItem]()
+    
+    var currentItems = [String : StatusItem]()
+
+    override var isFlipped: Bool {
+        return true;
+    }
+
+    func setup(_ editView: NSView) {
+        self.translatesAutoresizingMaskIntoConstraints = false
+        self.heightAnchor.constraint(equalToConstant: statusBarHeight).isActive = true
+        self.widthAnchor.constraint(equalTo: editView.widthAnchor).isActive = true
+        self.leadingAnchor.constraint(equalTo: editView.leadingAnchor).isActive = true
+        self.trailingAnchor.constraint(equalTo: editView.trailingAnchor).isActive = true
+        self.bottomAnchor.constraint(equalTo: editView.bottomAnchor).isActive = true
+    }
+
+    // Adds a status bar item. Only appends status bar items for now.
+    func addStatusItem(_ item: StatusItem) {
+        // If item exists, update its value
+        if let existingItem = currentItems[item.key] {
```

Comment:
I think trying to add an item that already exists either replaces the exiting item or is an error. Having it just silently update instead seems like an easy way to let people write bugs.